Zend/zend_opcode.c

```diff
@@ -599,9 +599,110 @@ static uint32_t zend_get_brk_cont_target(const zend_op_array *op_array, const ze
+		init->op1_type = IS_TMP_VAR;
+		return ZEND_CALL_FRAME_SLOT;
+	}
+
+	live_ranges = emalloc(sizeof(uint32_t) * num_args);
+
+	if (init->opcode == ZEND_INIT_FCALL && init->result.num == 0) {
+		zend_function *fbc = zend_hash_find_ptr(CG(function_table), Z_STR_P(CT_CONSTANT_EX(op_array, init->op2.constant)));
+		ZEND_ASSERT(fbc);
+		if (ZEND_USER_CODE(fbc->type)) {
+			num_args = MAX(num_args, fbc->op_array.num_args);
+		}
+	} else {
+		num_args += ZEND_RESERVED_RECV_SLOTS;
+	}
+
+	if (init->opcode != ZEND_NEW) {
+		/* only one result var per opcode ... can't set object *and* execute_data as result ... maybe find a better solution? */
+		init->result.var = offset + num_args;
+		init->result_type = IS_TMP_VAR;
+	}
+
+	while (1) {
+		switch ((++*opline)->opcode) {
+			case ZEND_NEW:
+			case ZEND_INIT_FCALL_BY_NAME:
+			case ZEND_INIT_FCALL:
+			case ZEND_INIT_NS_FCALL_BY_NAME:
+			case ZEND_INIT_METHOD_CALL:
+			case ZEND_INIT_STATIC_METHOD_CALL:
+			case ZEND_INIT_USER_CALL:
+			case ZEND_INIT_DYNAMIC_CALL:
+			case ZEND_INCLUDE_OR_EVAL:
+				fcall_Ts = zend_resolve_fcall(op_array, opline, offset + num_args + ZEND_CALL_FRAME_SLOT);
+				if (fcall_Ts > max_fcall_Ts) {
+					max_fcall_Ts = fcall_Ts;
+				}
+				break;
+
+			case ZEND_DO_FCALL:
+			case ZEND_DO_ICALL:
+			case ZEND_DO_UCALL:
+			case ZEND_DO_FCALL_BY_NAME:
+			case ZEND_DO_UNPACK_FCALL:
+				end = *opline - op_array->opcodes;
+				while (lr_off--) {
+					zend_live_range *range = op_array->live_range + live_ranges[lr_off];
+					if (range->start == end && live_ranges[lr_off] == op_array->last_live_range - 1) {
+						op_array->last_live_range--;
+					} else {
+						range->end = end;
+					}
+				}
+				zend_end_live_range(op_array, (*opline)->op1.num, *opline - op_array->opcodes, ZEND_LIVE_EXECUTE_DATA, offset + num_args);
+				(*opline)->op1_type = IS_TMP_VAR;
+				(*opline)->op1.var = offset + num_args;
+
+				efree(live_ranges);
+				return ZEND_CALL_FRAME_SLOT + num_args + max_fcall_Ts;
+
+			case ZEND_SEND_VAL:
+			case ZEND_SEND_VAR_EX:
+			case ZEND_SEND_REF:
+			case ZEND_SEND_VAR_NO_REF:
+			case ZEND_SEND_VAL_EX:
+			case ZEND_SEND_VAR:
+			case ZEND_SEND_USER:
+				live_ranges[lr_off] = (*opline)->result.num;
+				(*opline)->result.var = num_args - ((*opline)->extended_value & ZEND_ARG_SEND_MASK) + offset;
+				(*opline)->result_type = IS_VAR;
+				op_array->live_range[live_ranges[lr_off]].var = ((*opline)->result.var * sizeof(zval)) | ZEND_LIVE_TMPVAR;
+				lr_off++;
+			case ZEND_SEND_UNPACK:
+			case ZEND_FETCH_FUNC_ARG:
+				(*opline)->op2.var = offset + num_args;
+				(*opline)->op2_type = IS_TMP_VAR;
+				break;
+
+			/* No operands left, using ZEND_OP_DATA here */
+			case ZEND_FETCH_STATIC_PROP_FUNC_ARG:
+			case ZEND_FETCH_DIM_FUNC_ARG:
+			case ZEND_FETCH_OBJ_FUNC_ARG:
+				(*opline)++;
+				ZEND_ASSERT((*opline)->opcode == ZEND_OP_DATA);
+				(*opline)->op2.var = offset + num_args;
+				(*opline)->op2_type = IS_TMP_VAR;
+				break;
+		}
+	}
+}
+
```

Comment:
At least now I got, why you do this in pass_two :)
Numbering during compilation and simple update at pass_two would be better.

On Wed, Dec 16, 2015 at 6:41 PM, Bob Weinand <notifications@github.com>
wrote:

> In Zend/zend_opcode.c
> <https://github.com/php/php-src/pull/1679#discussion_r47790937>:
>
> > +				(*opline)->op2_type = IS_TMP_VAR;
> > +				break;
> > +
> > +			/* No operands left, using ZEND_OP_DATA here */
> > +			case ZEND_FETCH_STATIC_PROP_FUNC_ARG:
> > +			case ZEND_FETCH_DIM_FUNC_ARG:
> > +			case ZEND_FETCH_OBJ_FUNC_ARG:
> > +				(*opline)++;
> > +				ZEND_ASSERT((*opline)->opcode == ZEND_OP_DATA);
> > +				(*opline)->op2.var = offset + num_args;
> > +				(*opline)->op2_type = IS_TMP_VAR;
> > +				break;
> > +		}
> > +	}
> > +}
> > +
>
> We can have a separate counter (like op_array->T) with a special type in
> the compiler and have it translated to temps with an offset of op_array->T
> in pass_two() (the assertion that fcall temps are at the end of the
> op_array still needs to hold)… would be a nice cleanup at the very end,
> when everything else is resolved.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/php/php-src/pull/1679/files#r47790937>.
>